File: benchmarks/comparison/struct_ops.rs
struct Vec3 { x: i64, y: i64, z: i64 }

fn dot(a: &Vec3, b: &Vec3) -> i64 {
    a.x * b.x + a.y * b.y + a.z * b.z
}

fn main() {
    let mut total: i64 = 0;
    for i in 0i64..10_000_000 {
        let a = Vec3 { x: i, y: i + 1, z: i + 2 };
        let b = Vec3 { x: i + 3, y: i + 4, z: i + 5 };
        total += dot(&a, &b);
    }
    println!("{}", total);
}

File: benchmarks/comparison/struct_ops.c
#include <stdio.h>
#include <stdint.h>

typedef struct { int64_t x, y, z; } Vec3;

static int64_t dot(Vec3 a, Vec3 b) {
    return a.x * b.x + a.y * b.y + a.z * b.z;
}

int main(void) {
    int64_t total = 0;
    for (int64_t i = 0; i < 800000000; i++) {
        Vec3 a = {i ^ total, i + 1, i + 2};
        Vec3 b = {i + 3, i + 4, i + 5};
        total += dot(a, b);
    }
    printf("%ld\n", total);
    return 0;
}
